data-processing/pulsar-flink/2.7.0/pulsar-flink.md

```diff
@@ -0,0 +1,536 @@
+             url 'https://dl.bintray.com/streamnative/maven'
+         }
+}
+```
+
+For gradle projects, you can use the following [shade](https://imperceptiblethoughts.com/shadow/) plugin definition template to build an application JAR package that contains all the dependencies required for the client library and Pulsar Flink connector.
+
+```groovy
+buildscript {
+     repositories {
+         jcenter()
+     }
+     dependencies {
+         classpath 'com.github.jengelman.gradle.plugins:shadow:6.0.0'
+     }
+}
+
+apply plugin: 'com.github.johnrengelman.shadow'
+apply plugin: 'java'
+```
+
+# Advanced features
+
+This section describes advanced features supported by Pulsar Flink connector.
+
+## Pulsar primitive types
+
+Pulsar provides some basic native types. To use these native types, you can support them in the following ways.
+
+### Stream API environment
+
+PulsarPrimitiveSchema is an implementation of the `PulsarDeserializationSchema` and `PulsarSerializationSchema` interfaces.
+
+You can create the required instance in a similar way `new PulsarSerializationSchema(String.class)`.
+
+### Table environment
+
+We have created a new Flink format component called `atomic` that you can use in SQL format. In Source, it translates the Pulsar native type into only one column of RowData. In Sink, it translates the first column of RowData into the Pulsar native type and writes it to Pulsar.
+
+## Upsert Pulsar
+
+There is an increasing demand for Upsert mode message queues for three main reasons.
+
+- Interpret the Pulsar topic as a changelog stream, which interprets records with keys as Upsert events.
+- As part of the real-time pipeline, multiple streams are connected for enrichment and the results are stored in the Pulsar topic for further computation. However, the results may contain updated events.
+- As part of the real-time pipeline, the data stream is aggregated and the results are stored in Pulsar Topic for further computation. However, the results may contain updated events.
+  
+Based on these requirements, we support Upsert Pulsar. With this feature, users can read data from and write data to Pulsar topics in an Upsert fashion.
+
+In the SQL DDL definition, you can set the connector to `upsert-pulsar` to use the Upsert Pulsar connector.
+
+In terms of configuration, the primary key of the Table must be specified, and `key.fields` cannot be used.
+
+As a source, the Upsert Pulsar connector produces changelog streams, where each data record represents an update or deletion event. More precisely, the value in a data record is interpreted as a UPDATE of the last value of the same key, if this key exists (If the corresponding key does not exist, the UPDATE is considered as an INSERT.). Using the table analogy, data records in the changelog stream are interpreted as UPSERT, also known as INSERT/UPDATE, because any existing row with the same key is overwritten. Also, a message with a null value is treated as a DELETE message.
```

Comment:
```suggestion
As a source, the Upsert Pulsar connector produces changelog streams, where each data record represents an update or deletion event. More precisely, the value in a data record is interpreted as an UPDATE of the last value of the same key, if this key exists (If the corresponding key does not exist, the UPDATE is considered as an INSERT.). Using the table analogy, data records in the changelog stream are interpreted as UPSERT, also known as INSERT/UPDATE, because any existing row with the same key is overwritten. Also, a message with a null value is treated as a DELETE message.
```